agent/uiserver/uiserver_test.go

```diff
@@ -0,0 +1,230 @@
+package uiserver
+
+import (
+	"encoding/json"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"regexp"
+	"testing"
+
+	"github.com/hashicorp/consul/agent/config"
+	"github.com/hashicorp/consul/sdk/testutil"
+	"github.com/stretchr/testify/require"
+)
+
+func TestUIServer(t *testing.T) {
+	cases := []struct {
+		name            string
+		cfg             *config.RuntimeConfig
+		path            string
+		wantStatus      int
+		wantContains    []string
+		wantNotContains []string
+		wantEnv         map[string]interface{}
+		wantUICfgJSON   string
+	}{
+		{
+			name:         "basic UI serving",
+			cfg:          basicUIEnabledConfig(),
+			path:         "/", // Note /index.html redirects to /
+			wantStatus:   http.StatusOK,
+			wantContains: []string{"<!-- CONSUL_VERSION:"},
+			wantEnv: map[string]interface{}{
+				"CONSUL_ACLS_ENABLED": false,
+			},
+		},
+		{
+			// TODO: is this really what we want? It's what we've always done but
+			// seems a bit odd to not do an actual 301 but instead serve the
+			// index.html from every path... It also breaks the UI probably.
```

Comment:
I'm guessing this is intentional because the UI is a single-page app that parses the path to decide what to render, like when navigating to `http://localhost:8500/ui/dc1/services` we don't actually want to redirect, but rather serve the UI code and let JavaScript control rendering the `dc1` services "page".

Reply:
Ah good point. I'll remove the TODO.